### ISSUE TITLE
Crash when AI is adding document to internal memory

### ISSUE THREAD

**Issue by artytc** [ccbogel/QualCoder#965]
**Describe the bug :**
<br/>
When opening a project with a document, Qualcoder crashes at the "adding document to internal memory" step.
When AI is disabled, opening a project would not crash.
There are no error/exception messages in the log nor the console.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Version 10.0.27729

**Comment by kaixxx** [comment]
Which version of python do you have installed? 
Could you post the results of `pip list`? Make sure to run it in the same virtual environment as QualCoder. I would like to take a look at the versions of the python packages on your system. 
We have seen a similar error one before. But it somehow disappeared before we could figure out the exact cause of the issue.

**Comment by artytc** [comment]
Python version is 3.12.7

pip list:
aiohappyeyeballs                         2.4.3
aiohttp                                  3.10.10
aiosignal                                1.3.1
annotated-types                          0.7.0
anyio                                    4.6.2.post1
asgiref                                  3.8.1
attrs                                    24.2.0
backoff                                  2.2.1
bcrypt                                   4.2.0
build                                    1.2.2.post1
cachetools                               5.5.0
certifi                                  2024.8.30
cffi                                     1.17.1
chardet                                  5.2.0
charset-normalizer                       3.4.0
chroma-hnswlib                           0.7.3
chromadb                                 0.5.0
click                                    8.1.7
colorama                                 0.4.6
coloredlogs                              15.0.1
cryptography                             43.0.3
dataclasses-json                         0.6.7
Deprecated                               1.2.14
diff-match-patch                         20241021
distro                                   1.9.0
durationpy                               0.9
EbookLib                                 0.18
et_xmlfile                               2.0.0
fastapi                                  0.115.4
filelock                                 3.16.1
flatbuffers                              24.3.25
frozenlist                               1.5.0
fsspec                                   2024.10.0
fuzzysearch                              0.7.3
google-auth                              2.36.0
googleapis-common-protos                 1.65.0
greenlet                                 3.1.1
grpcio                                   1.67.1
h11                                      0.14.0
httpcore                                 1.0.6
httptools                                0.6.4
httpx                                    0.27.2
httpx-sse                                0.4.0
huggingface-hub                          0.26.2
humanfriendly                            10.0
idna                                     3.10
importlib_metadata                       8.5.0
importlib_resources                      6.4.5
Jinja2                                   3.1.4
jiter                                    0.7.0
joblib                                   1.4.2
json_repair                              0.30.1
jsonpatch                                1.33
jsonpointer                              3.0.0
kubernetes                               31.0.0
langchain                                0.3.7
langchain-chroma                         0.1.4
langchain-community                      0.3.5
langchain-core                           0.3.15
langchain-openai                         0.2.6
langchain-text-splitters                 0.3.2
langsmith                                0.1.142
lxml                                     5.3.0
markdown-it-py                           3.0.0
MarkupSafe                               3.0.2
marshmallow                              3.23.1
mdurl                                    0.1.2
mmh3                                     5.0.1
monotonic                                1.6
mpmath                                   1.3.0
multidict                                6.1.0
mypy-extensions                          1.0.0
networkx                                 3.4.2
numpy                                    1.26.4
oauthlib                                 3.2.2
onnxruntime                              1.20.0
openai                                   1.54.3
openpyxl                                 3.1.5
opentelemetry-api                        1.28.1
opentelemetry-exporter-otlp-proto-common 1.28.1
opentelemetry-exporter-otlp-proto-grpc   1.28.1
opentelemetry-instrumentation            0.49b1
opentelemetry-instrumentation-asgi       0.49b1
opentelemetry-instrumentation-fastapi    0.49b1
opentelemetry-proto                      1.28.1
opentelemetry-sdk                        1.28.1
opentelemetry-semantic-conventions       0.49b1
opentelemetry-util-http                  0.49b1
orjson                                   3.10.11
overrides                                7.7.0
packaging                                24.2
pandas                                   2.2.3
pdfminer.six                             20240706
pillow                                   11.0.0
pip                                      24.3.1
plotly                                   5.24.1
ply                                      3.11
posthog                                  3.7.0
propcache                                0.2.0
protobuf                                 5.28.3
pyasn1                                   0.6.1
pyasn1_modules                           0.4.1
pycparser                                2.22
pydantic                                 2.9.2
pydantic_core                            2.23.4
pydantic-settings                        2.6.1
Pygments                                 2.18.0
PyPika                                   0.48.9
pyproject_hooks                          1.2.0
PyQt6                                    6.7.1
PyQt6-Qt6                                6.7.3
PyQt6_sip                                13.8.0
pyreadline3                              3.5.4
python-dateutil                          2.9.0.post0
python-dotenv                            1.0.1
python-vlc                               3.0.21203
pytz                                     2024.2
PyYAML                                   6.0.2
QtAwesome                                1.3.1
QtPy                                     2.4.2
regex                                    2024.11.6
requests                                 2.32.3
requests-oauthlib                        2.0.0
requests-toolbelt                        1.0.0
rich                                     13.9.4
rispy                                    0.9.0
rsa                                      4.9
safetensors                              0.4.5
scikit-learn                             1.5.2
scipy                                    1.14.1
sentence-transformers                    3.2.1
setuptools                               75.3.0
shellingham                              1.5.4
six                                      1.16.0
sniffio                                  1.3.1
SQLAlchemy                               2.0.35
starlette                                0.41.2
sympy                                    1.13.1
tenacity                                 9.0.0
threadpoolctl                            3.5.0
tiktoken                                 0.8.0
tokenizers                               0.20.3
torch                                    2.5.1
tqdm                                     4.67.0
transformers                             4.46.2
typer                                    0.13.0
typing_extensions                        4.12.2
typing-inspect                           0.9.0
tzdata                                   2024.2
urllib3                                  2.2.3
uvicorn                                  0.32.0
watchfiles                               0.24.0
websocket-client                         1.8.0
websockets                               14.0
wrapt                                    1.16.0
yarl                                     1.17.1
zipp                                     3.20.2

**Comment by kaixxx** [comment]
Thank you. I checked the list of packages, everything looks fine. I even tried switching to python 3.12.7 to see if this is the problem - no, works fine on my machine. 

Last time I saw a similar issue, updating to the latest code solved it. Can you try that? 

If this does not help, some more ideas: 
- is the error happening directly after the message "adding document to internal memory..." appears in the action log, or after a while?
- Please try creating a new empty project, then enable the AI and add a simple and short text file as a document. This should rule out any issues with special file types. 

It's hard to track down the issue without a proper error message, but we will see.

**Comment by artytc** [comment]
I did a clean reinstall and this time it gets through the "adding document to internal memory" part. But when I perform an AI code search, it freezes after a few seconds and crash (similar to the first problem)

**Comment by kaixxx** [comment]
@artytc Thank you very much for investigating this further, that's a big help. I really want to get this problem sorted before the official release of version 3.6.

I am pretty sure it is an issue with chromadb, the database that we use as the 'local memory' of the AI. There is a known problem with chromadb on Windows, causing similar symptoms. We are using an older version of chromadb which should not be affected. But people also reported that they had to **downgrade to Python 3.10.** Could you try that?
 
Please ask back if you need any help switching python versions. You probably also need to reinstall requirements.txt to adapt everything to the older version of python. 

Thanks!

The chromadb issue: https://github.com/chroma-core/chroma/issues/2513

**Comment by kaixxx** [comment]
More as a reminder for me - a second report about the issue with chromadb, this time when querying: https://github.com/chroma-core/chroma/issues/3058
That's probably what happens now when you start the "AI Search".

**Comment by artytc** [comment]
I just downgraded to **Python 3.10.11**, reinstalled Qualcoder and dependencies, and the issue is gone.
I have also installed Qualcoder with Python 3.12.7 on my Mac and opened the same project, and there is no issue with AI search. Therefore, it should be a Windows-specific problem.

Just FYI, My documents are in Traditional Chinese. I am not sure if this contributes to the issue.

**Comment by kaixxx** [comment]
That's great, thank you very much. This confirms that we should stick with python 3.10 on Windows for now. 

Traditional Chinese should not cause any problems. I am using a multilingual embedding model (https://huggingface.co/intfloat/multilingual-e5-large). Traditional Chinese is part of the training data set, but might not be supported as well as some other languages. Would be interesting to know how it goes. 

FYI: I have added some information about the AI based functions to the wiki over the last couple of days, see:
- https://github.com/ccbogel/QualCoder/wiki/2.3.-AI-Setup
- https://github.com/ccbogel/QualCoder/wiki/4.2.-AI-Assisted-Coding
- https://github.com/ccbogel/QualCoder/wiki/5.1.-AI-chat-based-analysis
